### PR TITLE
Add support for arbitrary kickstart file injection into ISOs (HMS-3879)

### PR DESCRIPTION
**Pull request by achilleas-k**
See https://github.com/osbuild/images/pull/631

### CODE REVIEW

**Comment by chunfuwen** [comment]
Does this PR aim to address issue Lhttps://github.com/osbuild/bootc-image-builder/issues/433?

**Comment by cdrage** [comment]
@achilleas-k I'm trying to figure out how to test this, I've got a cluster machine that uses RAID5 so I'm hoping to inject something like this:

```
# Use the entire disk automatically
clearpart --all --initlabel

# Create RAID partitions on each disk
part raid.01 --size=1 --ondisk=sda --asprimary
part raid.02 --size=1 --ondisk=sdb --asprimary
part raid.03 --size=1 --ondisk=sdc --asprimary
part raid.04 --size=1 --ondisk=sdd --asprimary

part raid.1 --size=100 --grow --ondisk=sda --asprimary
part raid.2 --size=100 --grow --ondisk=sdb --asprimary
part raid.3 --size=100 --grow --ondisk=sdc --asprimary
part raid.4 --size=100 --grow --ondisk=sdd --asprimary

# Create the RAID 5 array
raid / --level=5 --device=md0 --fstype="ext4" raid.1 raid.2 raid.3 raid.4

# Specify bootloader installation (adjust as necessary for your setup)
bootloader --location=mbr --driveorder=sda,sdb,sdc,sdd --append="rhgb quiet"
```

into it, then boot it and have it auto-configure.

For this PR, I'm unable to see what volume / param I should pass in to `bootc-image-builder` image to pass in the kickstart file?

My second question is that I'm assuming that I won't need to elaborate on the entire install in the kickstart file / just inject what I need?

**Comment by achilleas-k** [comment]
I haven't tested it here yet, though it's the same code as in `obuild/images` which has been tested.  The way to add it is through the config.toml (toml is more convenient than json in this case):
```toml
[customizations.installer.kickstart]
contents = """
<bunch of kickstart stuff>
"""
```

> My second question is that I'm assuming that I won't need to elaborate on the entire install in the kickstart file / just inject what I need?

bootc-image-builder will add the `ostreecontainer` line for the base container and then append the stuff from customizations and that's it.  Everything else is up to the user.

**Comment by chunfuwen** [comment]
@achilleas-k  So when pass in user own kickstart file to bootc-image-builder ,  what happen to previously existed  kickstart for unattended iso ?  Would both kickstart files be merged as one ?

**Comment by achilleas-k** [comment]
> @achilleas-k So when pass in user own kickstart file to bootc-image-builder , what happen to previously existed kickstart for unattended iso ? Would both kickstart files be merged as one ?

bootc-image-builder will add the `ostreecontainer` line for the base container and then append the stuff from customizations and nothing else.
It's up to the user to write a kickstart that makes the installation unattended, interactive, or anything they want.  Doing anything more would be incredibly error prone.  We would have to parse the user's kickstart contents to figure out what we need to add (and where) and that's far beyond the scope of this change.

**Comment by achilleas-k** [comment]
Tests added.  Ready for review.

**Review by ondrejbudai — COMMENTED** [review]
This is awesome, can you please document it in `README.md`?

**Comment by ondrejbudai** [comment]
Also, one usability thought: IIUC, the kickstart is now passed via customization.toml... I wonder if we should also allow just mounting it in, e.g. `-v kickstart.ks:/kickstart.ks`. This way, users wouldn't have to embed their already existing kickstarts into a blueprint.

**Comment by achilleas-k** [comment]
> This is awesome, can you please document it in `README.md`?

Done!

**Comment by achilleas-k** [comment]
> Also, one usability thought: IIUC, the kickstart is now passed via customization.toml... I wonder if we should also allow just mounting it in, e.g. `-v kickstart.ks:/kickstart.ks`. This way, users wouldn't have to embed their already existing kickstarts into a blueprint.

Not sure how much easier that makes it, but if you want we can add it.  Here or follow-up?

**Review by mvo5 — COMMENTED** [review]
Thank you! This is nice and I'm happy to approve, but I do have a question/discussion point inline that I would like to resolve first (could be resolved in a followup of course if we are in a hurray :)

File: README.md [hunk 1]
```diff
@@ -332,6 +332,32 @@ Example:
...
+    "installer": {
+      "kickstart": {
+        "contents": "text --non-interactive\nzerombr\nclearpart --all --initlabel --disklabel=gpt\nautopart --noswap --type=lvm\nnetwork --bootproto=dhcp --device=link --activate --onboot=on"
+      }
+    }
+  }
+}
+```
+
+Since multi-line strings are difficult to write and read in json, it's easier to use the toml format when adding kickstart contents:
```
[the remark this answers is not in the document]
Comment (ondrejbudai) [reply README.md:349]: I agree! :)

File: test/test_manifest.py [hunk 2]
```diff
@@ -227,6 +248,66 @@ def test_manifest_user_customizations_toml(tmp_path, build_container):
...
+    # expected values for the following inline file contents
+    #
+    # %include /run/install/repo/osbuild-base.ks
+    # autopart --type=lvm
+    expected_content_hash = "dd63f9315bb74889b2b1a23c4f283571db122f2b11cf3161ef67cea3280f78b0"
+    expected_content_id = f"sha256:{expected_content_hash}"   # hash with algo prefix
+    expected_data = "JWluY2x1ZGUgL3J1bi9pbnN0YWxsL3JlcG8vb3NidWlsZC1iYXNlLmtzCmF1dG9wYXJ0IC0tdHlwZT1sdm0K"
+
+    # check the inline source for the custom kickstart contents
+    assert expected_content_id in manifest["sources"]["org.osbuild.inline"]["items"]
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply test/test_manifest.py:281]: You're probably right.  This is basically testing osbuild/images.  A pipeline name change shouldn't cause a test in a downstream project to fail, so if we change `bootiso-tree` for example, it'd be weird to have to adjust tests here (but probably okay to adjust tests in osbuild/images).

I'll keep the check for `sources."org.osbuild.inline".items` and leave it there.  I think that test is a good way of checking whether the kickstart content made it into the manifest.  It checks that the configuration is properly "wired up" to the manifest generation and the rest (copy stages and such) is the business of the images library, which already has tests that custom kickstart content ends up in the right places.

File: test/test_manifest.py [hunk 3]
```diff
@@ -227,6 +248,66 @@ def test_manifest_user_customizations_toml(tmp_path, build_container):
...
+        "--privileged",
+        "-v", "/var/lib/containers/storage:/var/lib/containers/storage",
+        "-v", f"{config_toml_path}:/config.toml",
+        "--security-opt", "label=type:unconfined_t",
+        f'--entrypoint=["/usr/bin/bootc-image-builder", "manifest", "--type=anaconda-iso", "{container_ref}"]',
+        build_container,
+    ])
+    manifest = json.loads(output)
+
+    # expected values for the following inline file contents
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply test/test_manifest.py:274]: Done.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (achilleas-k) [reply README.md]: Done.

**Comment by achilleas-k** [comment]
Please re ~review~ nitpick :)

**Review by ondrejbudai — APPROVED** [review]
Thanks!

**Pull request merged**